infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAssetDao.kt

```diff
@@ -245,48 +248,96 @@ abstract class LayoutAssetDao<T : LayoutAsset<T>>(
+            mapOf(
                 "ids" to ids.distinct().map { it.intValue },
                 "publication_state" to layoutContext.state.name,
                 "design_id" to layoutContext.branch.designId?.intValue
-            ), ::toRowVersion
+            ),
+            ::toRowVersion,
         )
     }
 
-    override fun fetchOfficialVersionAtMomentOrThrow(id: IntId<T>, moment: Instant): RowVersion<T> =
-        fetchOfficialVersionAtMoment(id, moment) ?: throw NoSuchEntityException(table.name, id)
+    override fun fetchOfficialVersionAtMomentOrThrow(
+        branch: LayoutBranch,
+        id: IntId<T>,
+        moment: Instant,
+    ): RowVersion<T> =
+        fetchOfficialVersionAtMoment(branch, id, moment) ?: throw NoSuchEntityException(table.name, id)
 
     //language=SQL
     private val officialVersionAtMomentSql = """
-        select
-          case when v.deleted then null else v.id end as id,
-          case when v.deleted then null else v.version end as version
-        from ${table.versionTable} v
-        where
-          v.id = :id
-          and v.change_time <= :moment
-          and not v.draft
-          and design_id is null
-        order by v.change_time desc
-        limit 1
+        with
+          main as (
+            select v.id, v.version, v.deleted, false as is_design
+              from ${table.versionTable} v
+              where
+                v.id = :id
+                and v.change_time <= :moment
+                and not v.draft
+                and design_id is null
+              order by v.change_time desc
+              limit 1
+          ),
+          design as (
+            select v.id, v.version, v.deleted, true as is_design
+              from ${table.versionTable} v
+              where
+                (v.official_row_id = :id or v.id = :id)
+                and v.change_time <= :moment
+                and not v.draft
+                and design_id = :design_id
+              order by v.change_time desc
+              limit 1
+          )
+        select id, version
+          from (select * from main union all select * from design) tmp
+          where deleted = false
+          order by (case when is_design then 0 else 1 end)
+          limit 1
     """.trimIndent()
 
-    override fun fetchOfficialVersionAtMoment(id: IntId<T>, moment: Instant): RowVersion<T>? {
+    override fun fetchOfficialVersionAtMoment(
+        branch: LayoutBranch,
+        id: IntId<T>,
+        moment: Instant,
+    ): RowVersion<T>? {
         logger.daoAccess(AccessType.VERSION_FETCH, LocationTrack::class, id)
         val params = mapOf(
+            "design_id" to branch.designId?.intValue,
             "id" to id.intValue,
             "moment" to Timestamp.from(moment),
         )
         return jdbcTemplate.queryOptional(officialVersionAtMomentSql, params, ::toRowVersion)
     }
 
+    @Transactional
+    override fun deleteRow(rowId: IntId<T>): DaoResponse<T> {
```

Comment:
Tässä uuden poistometodin toteutus uutena myös

Reply:
Tämä pisti miettimään (varmasti tästä on puhuttu, mutta muistini on huono mutta lyhyt,) missäs tilanteissa on ok poistaa design-official? 🤔 Esimerkiksi siinä tilanteessa että me luodaan joku käsite designissa ja julkaistaan se sinne viralliseksi versioksi, niin sehän on main-mergeensä saakka for all intents and purposes se virallisin versio. Onko se OK, että sen voi vaan käydä nukettamassa jossain tilanteessa?

Reply:
Joo tämä on hiukan hazardi, mutta tällä mallin toiminnalla vähän pakollinen. Tuo main-julkaisu on ainoa paikka jossa official designin saa poistaa, mutta siinä se täytyy poistaa joten poiston pitää olla mahdollista. Käytännössä tuossa kohtaa pitää myös ratkoon viedä päivitetty versio suunnitelmasta, jossa kyseinen rivi on merkitty "COMPLETED" tilaan. Sitä osuutta ei toki vielä ole toteutettu.

Detaljimmin: Koska viralliset ID:t on tarkoitus säilyä, design-official poistuu väkisin siinä kohtaa kun se julkaistaan uutena oliona mainiin (koska muuten official ID pitäisi vaihtaa sillä designin ID on jo käytössä). Tuon takia designeille on otettu sama filosofia kun drafteille: ne poistuu main-julkaisussa aina (myös silloin kun virallinen ID on jo main-oliolla, jolloin ei sinällään olisi aivan pakko). Koska tuo 1. case tarkoittaa että rivi vain päivitetään main-officialiksi, ainoa tilanne kun tätä kutsutaan on tuo 2. case.

Reply:
Niin ja vähän auki taitaa vielä olla mitä tapahtuu suunnitelman perumisessa. Oma intuitio sanoisi että senkin pitäisi toimia kuten draftin revert, eli poistaa päätaulusta nuo oliot. Mutta siitähän keskusteltiin enemmänkin joku viikko sitten ja se on vähän ehkä kyseenalaisempi.

Ja lisänä vielä, jos tähän dao-metodiin keksii jonkun kivan guardin että väärinkäyttö ei olisi mahdollista, niin se olisi toki bueno myös.